New pdf generator for tickets

CristianDeluxe opened the pull request:
Hello, I have completely modified the way to generate Tickets in PDF.

Now they are generated using Intervention\Image, which allows to have a much nicer and safer design, since PDF generators do not usually parse CSS well and it takes a lot of hours to get an acceptable design in HTML/CSS that is supported by the parser, which can "break" any day.

Generating images ensures that they will always look the same and will not break the design from one version to another, even using other generators in the future.

- Deleted unmaintained package from 2017 nitmedia/wkhtml2pdf
- The PDFs of the tickets are now generated by the new barryvdh/laravel-dompdf, no more permission errors or necessary binaries.
- Deleted unnecessary code, I have avoided duplicating code as much as possible
- **Includes Tests!**

Here is the new default ticket design:

![image](https://user-images.githubusercontent.com/156531/69511776-cc1d8900-0f41-11ea-8658-029f2f2b5ba9.png)

With a flyer (event image)

![image](https://user-images.githubusercontent.com/156531/69511802-e5263a00-0f41-11ea-92e1-5f9adce0f0b9.png)

I suggest to make a new field in the database/form with an exclusive image for the ticket (flyer_image?), which would allow graphic designers to adapt better to it, the recommended size would be 1360x635, but this can be done in future PRs, now adapts the event image.

Tell me what you think @etiennemarais :)

Leen15 committed:
Add docker production environment

Leen15 committed:
fix typos

etiennemarais committed:
📦 NEW: Add business fields to the confirmation display screen and email

etiennemarais committed:
📦 NEW: Add toggle functionality to the is business check. Test if the input sends data to the back.

bagage committed:
Do not escape html for post order message

etiennemarais committed:
Merge master

CristianDeluxe committed:
New PDF Generator

CristianDeluxe committed:
New PDF Generator

CristianDeluxe committed:
New PDF generator

CristianDeluxe committed:
Generate PDF name correctly

CristianDeluxe committed:
Remove unused code

CristianDeluxe committed:
Restore default value

CristianDeluxe committed:
Remove unused CSS

CristianDeluxe committed:
Update Language Files

CristianDeluxe committed:
Correct comments

CristianDeluxe committed:
Remove generatePdfTickets function

Not used

CristianDeluxe committed:
Add Laravel Debug Bar to composer.json

CristianDeluxe committed:
Remove debugbar from app.php as you don't need to define from Laravel 5.5>.

CristianDeluxe committed:
Use new PDF name generator

CristianDeluxe committed:
Remove unused files

CristianDeluxe committed:
Updated composer.lock

CristianDeluxe committed:
try to minimize formatting changes for a better diff

CristianDeluxe committed:
New Ticket Generator

CristianDeluxe committed:
Merge remote-tracking branch 'Attendize/next' into new-pdf-generator-for-tickets-next

CristianDeluxe committed:
Merge remote-tracking branch 'Attendize/next' into new-pdf-generator-for-tickets-next

# Conflicts:
#	app/Models/Ticket.php

CristianDeluxe committed:
Reversing the changes lost when merging

publicarray committed:
Fix XSS and 500 error if version check fails

Poper123 commented:
Many thanks and very nice. 

Oh, I found little spelling mistakes:

resources/views/Public/ViewEvent/Partials/EventHiddenTicketsSelection.blade.php 
ticket_status => ticket.status

resources/views/Public/ViewEvent/Partials/EventTicketsSection.blade.php
ticket_status => ticket.status

app/Models/Ticket.php
ticket_status => ticket.status

clawfire commented:
Wow I LOVE the design proposal. Can this be merged in the next release? ❤

publicarray committed:
Slightly increase PDF contrast

publicarray committed:
Small improvements to PDF generator

* higher res image
* hide overflowing text, allow long text in PDF ticket
* fix custom background colour
* cleanup

samdb committed:
Merge branch 'master' into version-check-fixes

johannac committed:
Merge branch 'develop' into fix-post-order-display-message-escaping

johannac committed:
Merge pull request #629 from bagage/fix-post-order-display-message-escaping

Do not escape html for post order message

johannac committed:
Merge branch 'develop' into version-check-fixes

johannac committed:
Add installed function back to utils to fix 500 error in tests

johannac committed:
Consolidate two getVersionInfo functions into 1

johannac committed:
Add unit test for new parse_version function

johannac committed:
Fix types in new parse_version function and expand regex a bit

johannac committed:
Change version check to use github version file instead of old attendize website

johannac committed:
Merge pull request #743 from publicarray/version-check-fixes

Fix XSS and 500 error if version check fails

chrisbrosnan committed:
Update ManageAccountController.php

Corrected 500 internal server error in accessing Account Settings in top right dropdown options, caused by Validator being used when already in use within the same controller (app/Http/Controllers/ManageAccouuntController.php). No problems seem to be caused by this fix.

johannac committed:
Merge pull request #789 from chrisbrosnan/develop

Update ManageAccountController.php

featurist-au commented:
Is there any update on when this suggested new pdf generator for tickets may be live?

publicarray committed:
fast PDF generation with dompdf

new minimum requirements: https://github.com/dompdf/dompdf#requirements

    PHP version 7.1 or higher
    DOM extension
    MBString extension
    php-font-lib
    php-svg-lib

publicarray committed:
Merge remote-tracking branch 'origin/new-pdf-generator-for-tickets-next'

Author: @CristianDeluxe

CristianDeluxe committed:
Merge remote-tracking branch 'PublicArray/new-pdf-ticket' into new-pdf-generator-for-tickets-next

# Conflicts:
#	app/Http/Controllers/EventAttendeesController.php
#	app/Http/Controllers/EventCheckoutController.php
#	app/Models/Ticket.php
#	app/Services/PDFGenerator/TicketGenerator.php
#	composer.json
#	composer.lock
#	config/app.php
#	config/dompdf.php
#	resources/views/ManageEvent/Customize.blade.php
#	resources/views/Public/ViewEvent/Partials/PDFTicket.blade.php
#	tests/Feature/PDFTickets/Controllers/EventCheckoutTicketTest.php

publicarray commented:
ping @johannac

justynpride commented:
Any update on whether this would get merged?

Fl0Cri commented on app/Jobs/GenerateTicket.php:
```diff
@@ -2,93 +2,54 @@
     {
-        return ($this->attendee_reference_index != null);
+        $this->pdf_file = TicketGenerator::createPDFTicket($this->order, $this->attendee);
+        $this->pdf_file->error ?: $this->fail();
```
Hi, just found this small mistake while trying to use this PR.
Actually it means that `GenerateTicket` will fail if `$this->pdf_file->error === false` which is probably not what was intended ;-)

jigzstar commented:
This looks very good! Great work.

johannac committed:
Merge pull request #990 from Attendize/develop

Release Candidate 2.4.0

johannac committed:
Merge branch 'develop'

CristianDeluxe committed:
Merge remote-tracking branch 'remote/master' into new-pdf-generator-for-tickets

# Conflicts:
#	.env.example
#	app/Handlers/QueueHandler.php
#	app/Http/Controllers/EventAttendeesController.php
#	app/Http/Controllers/EventCheckoutController.php
#	app/Http/Controllers/EventOrdersController.php
#	app/Http/Controllers/MyBaseController.php
#	app/Jobs/GenerateTicketsJobBase.php
#	app/Mailers/AttendeeMailer.php
#	app/Mailers/OrderMailer.php
#	composer.json
#	composer.lock
#	config/app.php
#	resources/views/ManageEvent/Partials/TicketDesignPreview.blade.php

CristianDeluxe committed:
composer.lock

CristianDeluxe committed:
Merge branch 'new-pdf-generator-for-tickets' into new-pdf-generator-for-tickets-next

# Conflicts:
#	.env.example
#	app/Http/Controllers/EventAttendeesController.php
#	app/Http/Controllers/EventCheckoutController.php
#	app/Http/Controllers/ManageAccountController.php
#	app/Jobs/GenerateTicketsJobBase.php
#	app/Jobs/SendAttendeeInvite.php
#	app/Jobs/SendAttendeeTicket.php
#	app/Jobs/SendOrderTickets.php
#	app/Listeners/OrderCompletedListener.php
#	app/Mailers/AttendeeMailer.php
#	app/Mailers/OrderMailer.php
#	composer.lock
#	config/app.php
#	config/dompdf.php
#	resources/views/ManageEvent/Customize.blade.php
#	resources/views/Public/ViewEvent/Partials/ExampleTicket.blade.php
#	resources/views/Public/ViewEvent/Partials/PDFTicket.blade.php
#	routes/web.php

rmwpl commented:
Hi @CristianDeluxe, just tried to use your branch, but I'm getting `[2022-04-18 18:34:39] local.ERROR: Class 'Services\PaymentGateway\Factory' not found {"userId":1,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Class 'Services\\PaymentGateway\\Factory' not found`
Can you help?

justynpride commented:
@CristianDeluxe I'm just trying this and get an error when submitting the order. On the checkout page it gives a Whoops message, but does send all the expected emails with the new ticket files. Checking the logs I get these entries: 

`
[2022-05-04 17:18:15] local.DEBUG: Queueing Order Notification Job  
[2022-05-04 17:18:17] local.DEBUG: Queueing Order Tickets Job  
[2022-05-04 17:18:17] local.INFO: Generating ticket for order: #5. Reference lKhEU45  
[2022-05-04 17:18:17] local.DEBUG: Ticket generated for order 5. PDF file: /home/bookinme/web/roles.book-in.me/public_html/public/user_content/pdf_tickets/lKhEU45.pdf  
[2022-05-04 17:18:19] local.DEBUG: Queueing Attendee Ticket Jobs  
[2022-05-04 17:18:19] local.INFO: Generating ticket for order: #5. Reference lKhEU45  
[2022-05-04 17:18:19] local.DEBUG: Use ticket from cache: /home/bookinme/web/roles.book-in.me/public_html/public/user_content/pdf_tickets/lKhEU45.pdf  
[2022-05-04 17:18:50] local.ERROR: Unable to open file for reading [/home/bookinme/web/roles.book-in.me/public_html/public/user_content/pdf_tickets/lKhEU45-1.pdf] {"userId":1,"exception":"[object] (Swift_IoException(code: 0): Unable to open file for reading [/home/bookinme/web/roles.book-in.me/public_html/public/user_content/pdf_tickets/lKhEU45-1.pdf] at /home/bookinme/web/roles.book-in.me/public_html/vendor/swiftmailer/swiftmailer/lib/classes/Swift/ByteStream/FileByteStream.php:131)
`

Am wondering if I've missed something?

justynpride commented:
> Hi @CristianDeluxe, just tried to use your branch, but I'm getting `[2022-04-18 18:34:39] local.ERROR: Class 'Services\PaymentGateway\Factory' not found {"userId":1,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Class 'Services\\PaymentGateway\\Factory' not found`
> Can you help?

@rmwpl did you resolve the error? Just encountered it.

justynpride commented:
Has anyone got this to work on the latest release 2.6.0?

CristianDeluxe commented:
Sorry, I've been quite busy lately with other projects.

The problem with this PR is that it has been open for a very, very long time (since 2019) and the code has been getting separated from the main branch.

I can work on it if @johannac or somebody can assure me that there is at least some intention to merge it with the project.

Otherwise it would be wasted hours :(

Thanks for wanting to use this feature.

justynpride commented:
@CristianDeluxe that makes complete sense, as time is very precious. I hope we can implement your design, as it is a really amazing feature.

camlafit commented:
Hello

Looks related also to #1087 issue

Neustradamus commented:
Dear @johannac, @egaudrain, @ichier, @jochemp264, @justynpride,

Any progress on this @CristianDeluxe PR?

justynpride commented:
There is a helpful discussion on this at https://github.com/Attendize/Attendize/pull/935. 

I've personally moved away from this project, and for any booking system needs use hi.events (https://github.com/HiEventsDev/hi.events). This is in active development and does everything Attendize does and more, and is built on the latest versions of Laravel and Php etc..

camlafit commented:
Hello

Thanks for this up :) 
I've forgotten this. For my part, as simple user I've moved to pretix service. Their pdf generator is  very useful.

Nice to discover a new service as hi.events

rootjack404 commented:
@Neustradamus 

If you really need to use more modern ticket design, just change the pdf generator to Weasyprint, it support most of the modern HTML/CSS element. It's easy to implement as well, just use the Native Weasyprint PHP driver, but it comes with risk, since you need to allow shell_exec in the PHP config ;) I myself already used it, and it does really good in performance.

CristianDeluxe commented:
Hi everyone, as I mentioned in a previous post, I understand that this project is no longer being maintained for various reasons.

On my side, we’re creating a new software solution since, unfortunately, HiEvents didn’t quite work for me. Its configuration was too complex, and it was difficult to adapt it to my needs.

We’ve developed something new from scratch, written in Laravel and React, and it’s ready and working. We hope to launch it as open source in the coming months. One feature I’ve always loved about Attendize was the 'unzip and run' approach, and we’re sticking to that concept to ensure it can be installed on any affordable hosting. Additionally, it will support plugins (like payment gateways), as in my country (Spain), we use Redsys—a payment gateway used by most banks with much lower fees than PayPal or Stripe, also some cool features like pixel events for tracking ads funnels, etc...

If you’re interested, please react to this comment, and I’ll notify you when we release the first version. For now, it’s already processing thousands of tickets for real clients without any issues; we just need to clean up the code related to our business and make it open-source ready.

I want to thank the Attendize team for their incredible work. This software has brought me much joy. Unfortunately, it would require a tremendous effort to get the project back on track due to how outdated the code is.

justynpride commented:
> I want to thank the Attendize team for their incredible work. This software has brought me much joy. Unfortunately, it would require a tremendous effort to get the project back on track due to how outdated the code is.

I agree, the software has been of great benefit, and everyone hard work has been appreciated. Thank you.
@CristianDeluxe please do let me know when your software is open-source ready.

RobertKernDevelopment commented:
@CristianDeluxe please feel free to let me know when your software is available as open-source. I was actually considering checking out HiEvents, but I think I'll wait until your software is released.

robemorata commented:
Hi @CristianDeluxe, I'm very interested in the software you've developed! Please feel free to contact me; I'd love to stay updated and explore it further. Great work

CristianDeluxe committed:
Update event controllers and ticket email handling
